src/api/firebase.js

```diff
@@ -173,3 +173,71 @@ export async function deleteItem() {
 	 * this function must accept!
 	 */
 }
+
+/**
+ * Compare purchase history of list items and sort items by three attributes:
+ * 1) Purchase status, 2) Urgency, and 3) Alphabetically
+ * add a property to each item indicating the urgency of the next purchase
+ * @param {object} data The raw list data to be sorted.
+ */
+export function comparePurchaseUrgency(data) {
+	// Create an empty, nested array for sorted data categories
+	const sortedData = [];
+	for (let i = 0; i < 6; i++) {
+		sortedData.push([]);
+	}
+
+	// Loop through all items and assign urgency based on days until next purchase
+	for (let item of data) {
+		// If the item has been purchased before, return days since last purchase
+		// If the item has never been purchased, return the days since item creation
+		const daysSincePurchase = item.dateLastPurchased
+			? getDaysBetweenDates(item.dateLastPurchased)
+			: getDaysBetweenDates(item.dateCreated);
+
+		// Assign urgency
+		if (daysSincePurchase >= 60) {
+			item.purchaseStatus = 'Inactive';
+			item.urgencyCategory = 4; // Inactive
+		} else {
+			const daysUntilPurchase = item.dateLastPurchased
+				? getDaysBetweenDates(item.dateNextPurchased)
+				: getDaysBetweenDates(undefined, item.dateNextPurchased);
+
+			item.purchaseStatus = 'Active';
+			item.days = daysUntilPurchase;
+
+			if (daysUntilPurchase < 0) {
+				item.urgencyCategory = 0; // Overdue
+			} else if (daysUntilPurchase <= 7) {
+				item.urgencyCategory = 1; // Soon
+			} else if (daysUntilPurchase > 7 && daysUntilPurchase < 30) {
+				item.urgencyCategory = 2; // Kind of Soon
+			} else if (daysUntilPurchase >= 30) {
+				item.urgencyCategory = 3; // Not Soon
+			} else {
+				item.urgencyCategory = 4; // Inactive
+			}
+		}
+	}
+
+	//Sort items, first by purchase status,
+	//then by urgency, and lastly alphabetically.
+	data.sort(
+		(a, b) =>
+			a.purchaseStatus.localeCompare(b.purchaseStatus) ||
+			a.urgencyCategory - b.urgencyCategory ||
+			a.name.localeCompare(b.name),
+	);
+
+	// Group items by urgency category
+	data.forEach((item) => {
+		if (item.isChecked) {
+			sortedData[5].push(item);
```

Comment:
another magic number, I'd like to see the  number come from `itemUrgencyCategory`, or if it get's converted to an array creating a method called `getCategoryIndex` or something like that.

Reply:
Unfortunately, I tackled these comments in order, so I'm not sure if I'm handling itemUrgencyCategory the way that you had envisioned. I moved the logic for handling purchased items into `getItemUrgency` in order to eliminate the conditional.